vllm/compilation/fusion_attn.py

```diff
@@ -0,0 +1,146 @@
+            output_view = attn_getitem
+
+            while is_func(get_only_user(output_view), RESHAPE_OP):
+                output_view = get_only_user(output_view)
+
+            # If this is not a quant node, exit
+            if not is_auto_func(get_only_user(output_view), QUANT_OP):
+                continue
+
+            quant_node: fx.Node = get_only_user(output_view)
+            quant_output: fx.Node = quant_node.kwargs["result"]
+            assert is_func(quant_output, EMPTY_OP)
+            quant_output_shape = quant_output.args[0]
+
+            # Extract pertinent attention information,
+            # assuming the output is allocated 2d and reshaped
+            layer_name = node.kwargs["layer_name"]
+            attn_output_view = node.kwargs["output"]
+            assert is_func(attn_output_view, RESHAPE_OP)
+            attn_output_buf = attn_output_view.args[0]
+            assert is_func(attn_output_buf, EMPTY_OP)
+
+            # Check that the attention layer impl supports quant fusion
+            attn_layer: Attention = self.static_fwd_context[layer_name]
+            quant_dtype = quant_output.kwargs['dtype']
+            # hardcoded because we only match the op for static/per-tensor
+            # (static_scaled_fp8_quant) for now
+            static, per_token = True, False
+            if not attn_layer.impl.fused_output_quant_supported(
+                    quant_dtype, static=static, per_token=per_token):
+                continue
+
+            # The output shapes and dtypes are:
+            # - attn_view: bf16[s0, n_heads, head_size]
+            # - attn_buf: bf16[s0, n_heads * head_size]
+            # - quant: fp8[s0, n_heads * head_size]
+
+            # START REWRITE
+            # After this point, all operations must succeed for the graph to
+            # remain valid. That includes modifications to attention layers.
+
+            # 1. Set scale on unified_attn
+            node.update_kwarg('output_scale', quant_node.kwargs['scale'])
+
+            # 2. Rewrite the initial buf alloc dtype
+            attn_output_buf.update_kwarg('dtype', quant_output.kwargs['dtype'])
+            assert attn_output_buf.args == quant_output.args
+            assert attn_output_buf.kwargs == quant_output.kwargs
+            # meta value fixed by FakeTensorUpdater
+
+            # 3. Update the autofunc node meta, as the FakeTensorUpdater does
+            # not touch them. To get both the shape and type correct, we have
+            # to create the type manually, which is a bit ugly.
+            old_fake = node.meta['val'][1]
+            new_fake_tensor = torch.empty_like(old_fake, dtype=quant_dtype)
+            node.meta['val'] = (None, new_fake_tensor)
+
+            # 4. Autofunc nodes don't get updated, so add the next node to the
+            # update list. Also remove the meta value as the FakeTensorUpdater
+            # has a bug where the type is not considered in FakeTensor equality.
```

Comment:
Could you clarify what the bug is? (Also, please submit a bug to pytorch if you haven't already)

Reply:
Yep, it's on my to-do list to submit the bugs here.